textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -0,0 +1,170 @@
+package edu.uci.ics.textdb.dataflow.regexmatch;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.StringJoiner;
+
+import edu.uci.ics.textdb.common.constants.DataConstants;
+
+
+public class GramBooleanQuery {
+	public enum QueryOp {
+		NONE, // doesn't match any string
+		ANY,  // matches any string
+		
+		AND,
+		OR
+	}
+	QueryOp operator;
+	List<String> operandList;
+	List<GramBooleanQuery> subQueryList;
+	
+	private int gramLength;
+	
+	/**
+	 * Constructs a GramBooleanQuery with default gram length 3. <br>
+	 * @param operator
+	 */
+	public GramBooleanQuery(QueryOp operator) {
+		this(operator, 3);
+	}
+	
+	public GramBooleanQuery(QueryOp operator, int gramLength) {
+		this.operator = operator;
+		operandList = new ArrayList<String>();
+		subQueryList = new ArrayList<GramBooleanQuery>();
+		this.gramLength = gramLength;
+	}
+	
+	/**
+	 * This return a GramBooleanQuery's hash code. <br>
+	 * It won't traverse the whole tree, instead, 
+	 * it only calculates the hashcode of direct leafs. <br>
+	 * 
+	 */
+	public int hashCode() {
+		int hashCode = operator.toString().hashCode();
+		for (String s : operandList) {
+			hashCode = hashCode ^ s.hashCode();
+		}
+		return hashCode;
+	}
+	
+	public boolean equals(Object compareTo) {
+		if (! (compareTo instanceof GramBooleanQuery)) {
+			return false;
+		}
+		
+		GramBooleanQuery query = (GramBooleanQuery) compareTo;
+		if (this.operator != query.operator
+			|| this.operandList.size() != query.operandList.size()
+			|| this.subQueryList.size() != query.subQueryList.size()) {
+			return false;
+		}
+		
+		Set<String> operandSet = new HashSet<String>(this.operandList);
+		if (!operandSet.equals(new HashSet<String>(query.operandList))) {
+			return false;
+		}
+		
+		Set<GramBooleanQuery> subQuerySet = new HashSet<GramBooleanQuery>(this.subQueryList);
+		if (!subQuerySet.equals(new HashSet<GramBooleanQuery>(query.subQueryList))) {
+			return false;
+		}
+		
+		return true;
+	}
+	
+	/**
+	 * This methods takes a list of strings and adds them to the query tree. <br>
+	 * For example, if the list is {abcd, wxyz}, then: <br>
+	 * trigrams({abcd, wxyz}) = trigrams(abcd) OR trigrams(wxyz) <br>
+	 * OR operator is assumed for a list of strings. <br>
+	 * @param list, a list of strings to be added into query.
+	 */
+	public void add(List<String> list) {
+		addOrNode(list);
```

Comment:
If "addOrNode" is only used here, there may not be a need to have a separate function.